### PR TITLE
Add colData and rowData to unfiltered sce object

### LINKED ISSUE #20: Modify generation of unfiltered output to include basic cell statistics in colData
We will want to add in `colData` to the unfiltered output. In order to do that we should modify `generate_unfiltered_rds.R` to add in the `colData` prior to writing the rds file.

### PR DESCRIPTION
**Pull request by allyhawkins**
Closes #20. This PR adds the calculation of the `colData` and `rowData` statistics to the unfiltered output within the `generate_unfiltered_sce.R` script. To do this for the RNA counts I am using `add_cell_mito_qc()` within `scpcaTools` for the per cell statistics, to get total UMI count, number of genes detected per cell, and the mitochondrial content per cell. I then also used `scater::addPerFeatureQC()` to add in the number of counts for each gene in the `rowData`. Within the nextflow workflow, the only modification was the addition of the mitochondrial file as a paramter in the config file and then as an input to the process. 

I also made the choice to NOT use the `miQC` option here since I believe we only want that additional column on the filtered output. Additionally when playing around with it, I noticed that miQC doesn't handle empty drops well and threw a few errors with a sample that hadn't gone through filtering with `emptyDrops` filtering, another reason we probably only want to use it on samples post filtering with `emptyDrops`. 

Where I was not 100% sure about what to do was with the CITE-seq counts data. For those objects, I thought it could be useful to at least have total counts per cell and total number of counts per antibody. To do that I added in `colData` using `scater::addPerCellQC()` and `rowData` using `scater::addPerFeatureQC()` before merging the experiments within the `generate_unfiltered_sce.R` script. However, the merge does not keep the `colData` and `rowData` for the alternative experiment. I think we could either grab the specific statistics that we are interested in and add it in as metadata for the `altExp` or alter the `merge_altexp()` function to keep the data that we need (but the second option would be outside the scope of this PR). 

I did test these changes with sample `SCPCR000006` and everything looks as it should for a sample with just RNA counts.

### CODE REVIEW

**Review by jashapiro — COMMENTED** [review]
This looks good. My main suggestion is to do with the question that you had about feature stats:

> Where I was not 100% sure about what to do was with the CITE-seq counts data. For those objects, I thought it could be useful to at least have total counts per cell and total number of counts per antibody. To do that I added in `colData` using `scater::addPerCellQC()` and `rowData` using `scater::addPerFeatureQC()` before merging the experiments within the `generate_unfiltered_sce.R` script. However, the merge does not keep the `colData` and `rowData` for the alternative experiment. I think we could either grab the specific statistics that we are interested in and add it in as metadata for the `altExp` or alter the `merge_altexp()` function to keep the data that we need (but the second option would be outside the scope of this PR).

If we hold off on calculating the statistics until we have the merged SCE (after the `if` block), `scater::addPerCellQC()` (and presumably `add_cell_mito_qc()`) will add `altexps_ALTNAME_sum`,  `altexps_ALTNAME_detected` and  `altexps_ALTNAME_percent` columns to the `colData` matrix. I'm not sure of the value of the last of these stats is, but seems fine to leave it. (`ALTNAME` here would be the name of the altExp, here whatever the value of `opt$feature_name` is.) 

However, the feature stats are not added by `scater::addPerFeatureQC()` so to add that we would need to add that data to the `altExp` manually. I have suggested a way to do that below.  We could add cell stats directly as `colData` to the `altExp` as well in the same manner, but that would be duplicative of the data already in the main table, and would only be useful if people are extracting the matrix separately, in which case they may  want to recalculate these stats (and doing so is not onerous).

We could also alter `scpcaTools` to keep stats already present. It will probably mean adding NAs to the table, because while count statistics can have a default of 0 for cells not observed, it is not clear that would be the correct value for every kind of stat that might be included. So that makes my inclination to leave the `colData` and `rowData` unpopulated (or rather, leave the current behavior of dropping those stats during the merge).

File: modules/generate-rds.nf [hunk 1]
```diff
@@ -63,9 +67,8 @@ workflow generate_rds {
   // generate rds files for RNA-only samples
   take: quant_channel
   main:
-    quant_channel \
-      | make_unfiltered_sce \
-      | filter_sce
+    make_unfiltered_sce(quant_channel, params.mito_file)
+    filter_sce(make_unfiltered_sce.out)
```
[the remark this answers is not in the document]
Comment (allyhawkins) [reply modules/generate-rds.nf:72]: Ah I can try these options. I wasn't quite sure the best way to modify it with the pipes to add in both inputs which is why I had switched it back. I'll test out the `&` operator and see what happens.

**Comment by allyhawkins** [comment]
> However, the feature stats are not added by `scater::addPerFeatureQC()` so to add that we would need to add that data to the `altExp` manually. I have suggested a way to do that below.

Thanks for the helpful suggestions! This seems to have done the trick and in testing it, I now can access the `rowData` separately for the main sce and then the altExp. Also by adding in the addition of calculating the statistics after merging the `colData` now has both the rna and altExp data combined. I definitely prefer this to altering the function within `scpcaTools`.

File: modules/generate-rds.nf (hunk 1 above)
[the remark this answers is not in the document]
Comment (allyhawkins) [reply modules/generate-rds.nf]: In testing this, it looks like nextflow doesn't like the `&` operator. I tried it as you suggested and with parenthesis around it, `(quant_channel & params.mito_file)` and both times nextflow yelled about not recognizing the `&` symbol. Using the combined version worked well so I went ahead and changed it to that.

[the remark this answers is not in the document]
Comment (jashapiro) [reply modules/generate-rds.nf]: I guess the only way to use it is when passing a channel to two different processes. Seems odd that there is no way to use pipe syntax to pass in more than one channel, but not really my design decision.

File: bin/generate_unfiltered_sce.R [hunk 2]
```diff
@@ -36,6 +36,12 @@ option_list <- list(
     opt_str = c("-u", "--unfiltered_file"),
     type = "character",
     help = "path to output unfiltered rds file. Must end in .rds"
+  ),
+  make_option(
+    opt_str = c("-m", "--mito_file"),
+    type = "character",
+    default = "",
```
Comment (jashapiro) [inline bin/generate_unfiltered_sce.R:43]: If we want this to be required, probably better to not have a default at all.  (`feature_dir` has a default so we can use it to test for whether there is anything there in the `if` statement. Maybe not the best way to handle that, but it works) 
```suggestion
```

File: bin/generate_unfiltered_sce.R [hunk 3]
```diff
@@ -51,6 +57,16 @@ if(!(stringr::str_ends(opt$unfiltered_file, ".rds"))){
   stop("unfiltered file name must end in .rds")
 }
 
+# check that mitochondrial gene list exists
+if(!file.exists(opt$mito_file)){
+  stop("Mitochondrial gene list is required using -m or --mito_file")
```
Comment (jashapiro) [inline bin/generate_unfiltered_sce.R:62]: If the mito file  is required by option parser, then we will want to change this text.
```suggestion
  stop("Mitochondrial gene list file not found.")
```

File: bin/generate_unfiltered_sce.R [hunk 4]
```diff
@@ -51,6 +57,16 @@ if(!(stringr::str_ends(opt$unfiltered_file, ".rds"))){
...
+}
+
+# read in mitochondrial gene list
+mito_genes <- readr::read_tsv(opt$mito_file, col_names = "gene_id") %>%
+  dplyr::pull("gene_id") %>%
+  unique()
```
Comment (jashapiro) [inline bin/generate_unfiltered_sce.R:68]: If this is just a file with one line per row, we can simplify this a bunch:
```suggestion
mito_genes <- unique(scan(opt$mito_file, what = "character"))
```

**Review by jashapiro — APPROVED** [review]
Looks good, but for a few little comments.

**Pull request merged**